File: solutions/cpp/allergies/1/allergies.cpp
#include "allergies.h"
#include <array>

namespace allergies {

const std::array<std::string, 8> all_allergens = {
    "eggs",
    "peanuts",
    "shellfish",
    "strawberries",
    "tomatoes",
    "chocolate",
    "pollen",
    "cats"
};
    
allergy_test::allergy_test(unsigned code) {
    for (unsigned i = 0; i < all_allergens.size(); ++i)
        if (code & (1 << i))
            allergens.insert(all_allergens[i]);
}

bool allergy_test::is_allergic_to(const std::string& allergen) const {
    return allergens.find(allergen) != allergens.end();    
}

}  // namespace allergies

File: solutions/prolog/allergies/3/allergies.pl
allergies(Score, Allergies) :-
    findall(X, allergic_to(Score, X), Allergies).

allergic_to(Score, Allergen) :-
    bit(Allergen, Bit),
    has_bit(Score, Bit).

bit(eggs, 0).
bit(peanuts, 1).
bit(shellfish, 2).
bit(strawberries, 3).
bit(tomatoes, 4).
bit(chocolate, 5).
bit(pollen, 6).
bit(cats, 7).

has_bit(Value, Bit) :- Value /\ 2^Bit =:= 2^Bit.
